Headers Performance Boost and Interface Simplification

Scottmitch opened the pull request:
Motivation:
A degradation in performance has been observed from the 4.0 branch as documented in https://github.com/netty/netty/issues/3962.

Modifications:
- Simplify Headers class hierarchy.
- Restore the DefaultHeaders to be based upon DefaultHttpHeaders from 4.0.
- Make various other modifications that are causing hot spots.

Result:
Performance is now on par with 4.0.

Scottmitch commented:
@normanmaurer - PTAL. Basically everything in the `HttpHeaders` / `DefaultHttpHeaders` is `deprecated`. Most `deprecated` items have alternatives to use, but some describe that future releases will be changing this method (rather than bringing *everything* back, and some type signatures in master conflict).

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/EmptyHttpHeaders.java:
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
```
2015 ?

Scottmitch replied:
yip. fixed.

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpHeaderUtilTest.java:
```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2013 The Netty Project
```
2015

Scottmitch replied:
done

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTestUtils.java:
```diff
@@ -0,0 +1,130 @@
+        FIVE("five", 5),
+        SIX_QUOTED("six,", 6),
+        SEVEN_QUOTED("seven; , GMT", 7),
+        EIGHT("eight", 8);
```
I think we usually use CamelCase

Scottmitch replied:
Done

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java:
```diff
@@ -211,54 +241,48 @@ public int size() {
     @Override
     public Set<T> names() {
-        return unmodifiableSet(map.keySet());
+        Set<T> names = new LinkedHashSet<T>(size());
```
I guess we could first check if its empty and if so just use Collections.emptySet() ?

Scottmitch replied:
Sure. Done.

Scottmitch commented:
@normanmaurer - Addressed comments. My thought is to get this PR to an acceptable state, cherry-pick it, and then discuss the deprecation changes with @trustin and @nmittler which may result in a followup PR. Sound good?

trustin commented:
SGTM!

normanmaurer commented:
Ship it



> Am 15.08.2015 um 02:14 schrieb Trustin Lee <notifications@github.com>:
> 
> SGTM!
> —
> Reply to this email directly or view it on GitHub.
>

Scottmitch commented:
Cherry-picked 4.1 (ba6ce54).

Thanks for review all!
